07-13-16-language-plugins.md

```diff
@@ -0,0 +1,97 @@
+This month we merged one of the most significant pull requests for the Tessel project: [adding Rust deployment support](https://github.com/tessel/t2-cli/pull/790). For the past three years we've focused exclusively on deploying JavaScript on microcontrollers. This is our first step investing in a new language, a new ecosystem, and a new community. However, the purpose of this post is not to dive into the benefits of Rust, but rather to showcase the ease of adding new language support to the Tessel CLI.
+
+Despite the significance of this addition to the CLI codebase, the amount of code we modified was relatively small. We simply added two main files (along with minor modifications to a handful of other tests): [one with the implementation](https://github.com/tessel/t2-cli/blob/b8cd1f34398c2d188fa188a216b4b9b4dd206dc4/lib/tessel/deployment/rust.js) and one with [unit tests](https://github.com/tessel/t2-cli/blob/b8cd1f34398c2d188fa188a216b4b9b4dd206dc4/test/unit/deployment/rust.js). The simplicity of the feature addition is a testament [@rwaldron](https://github.com/rwaldron)'s hard work building the CLI's language plugin system. In this post, I want to demonstrate how straight forward it is to add support for a new language to the CLI using this plugin system.
+
+The process for deploying an arbitrary code bundle, regardless of language, to Tessel can be generalized. There are six functions that need to be exported for any language deployment mechanism:
+
+1. Determine Project Location in the file system
+2. Run any compilation/pre-processing steps (optional)
+3. Bundle the executable code into a tarball
+4. Run any pre-execution steps (optional)
+5. Execute the code/binary
+6. Run any post-execution steps (optional)
+
+These functions are more concisely named `checkConfiguration`, `preBundle`, `tarBundle`, `preRun`, `shell`/`binary`, and `postRun`. Here is an explanation of each of those interfaces.
+
+### 1. checkConfiguration
+`checkConfiguration` should be able to accept the directory that was deployed with `t2 run ...` and return the path to the entrypoint of the project. With Rust, we do this by parsing the `Cargo.toml` and with JavaScript we parse the project's `package.json`. It is expected to return an object with a `basename` and optional `program` property (for the name of the project).
+
+```.js
+// checkConfiguration to confirm project deployment files
+exportables.checkConfiguration = (pushdir, basename, program) => {
+  // Should return the details about the project
+  return {
+    basename,
+    program
+  };
+};
+```
+
+### 2. preBundle
+`preBundle` is an optional step for any processes that need to be run prior to the next step where you actually package the directory up to be sent to the Tessel 2. In JavaScript is integral to the binary compilation stage (where we inject our pre-compiled binaries in place of existing project binary dependencies) but it's less critical for Rust where we just set some optional parameters.
```

Comment:
Clearer: "For example, when deploying JavaScript code, this step is integral ... because ... In Rust, this step is less critical (it is possible to deploy without it), but we use it to ..."